### PR TITLE
Post for C/C++ issues on CRAN

### PR DESCRIPTION
**Pull request by ateucher**
@gaborcsardi and @DavisVaughan both had a look at an earlier draft.

### CODE REVIEW

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (jennybc) [inline content/blog/cran-checks-compiled-code/index.Rmd:126]: Maybe also explain the return value of `snprintf()`? Since it does come up below.

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (jennybc) [inline content/blog/cran-checks-compiled-code/index.Rmd:148]: I had cases that were harder to analyze, but I don't have anything productive to say other than "sometimes it's kind of hard to figure this out" 😬

**Review by jennybc — APPROVED** [review]
A few comments, but in general LGTM!

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [inline content/blog/cran-checks-compiled-code/index.Rmd:140]: This is probably overkill

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [inline content/blog/cran-checks-compiled-code/index.Rmd:205]: Generally, however, you can explain this problem when submitting your package, and as long as you've have notified the upstream maintainer, CRAN will accept your updated package.

**Review by hadley — APPROVED** [review]
Looks great!

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (tracykteal) [reply content/blog/cran-checks-compiled-code/index.Rmd:37]: I'd also suggest 'best practices' -> 'shared practices'

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (gaborcsardi) [reply content/blog/cran-checks-compiled-code/index.Rmd:75]: The date of 4.3.0 is not known yet.

**Review by tracykteal — APPROVED** [review]
Looks good to me too!

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd:5]: May as well be consistent. Thanks!

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd:148]: Do you have any examples you could point me to that might help me come up with a better example as per @hadley's [comment](https://github.com/tidyverse/tidyverse.org/pull/630#discussion_r1135822319)?

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd:140]: Do you mean the "Truncated..." Rprintf output?

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (jennybc) [reply content/blog/cran-checks-compiled-code/index.Rmd:151]: I don't think you can use this directly, but it's an example of a more complicated situation. Here's a patch in the grisu code that is vendored into readr and vroom.

https://github.com/tidyverse/readr/commit/9143718735136c05c52c467b996dff799f77b888

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd]: I pushed a new example... don't know if it's too contrived?

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (jennybc) [reply content/blog/cran-checks-compiled-code/index.Rmd]: And here's where it intersects with code we control:

https://github.com/tidyverse/readr/blob/9558d899b7e8c592be339bfc31ec35593ccf20ca/src/write_delim.cpp#L219

So I guess I could have hardwired a size of 32, but that would have been really opaque on the grisu side.

Comment (jennybc) [reply content/blog/cran-checks-compiled-code/index.Rmd]: This is the problem with these mandates, because the grisu function should really be refactored to "know" about the size of buffer that's available to it. But with vendored code, you want to make the smallest patch possible.

Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd]: That's great, thanks! I think it's a great example because you do a nice job explaining the choices you made. I did also just push up a new example, but I think it's pretty contrived. I could link to this instead (or as well)?

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [reply content/blog/cran-checks-compiled-code/index.Rmd:139]: Yeah

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [inline content/blog/cran-checks-compiled-code/index.Rmd:74]:
```suggestion
The tidyverse has a [policy of supporting four previous versions](https://www.tidyverse.org/blog/2019/04/r-version-support/) of R. Currently that includes R 3.5, but with the upcoming release of R 4.3 (which should be this Spring some time) the minimum version we will support is R 3.6. As we won't be supporting R 3.5 in the near future, you should not feel pressured to either.
```

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [inline content/blog/cran-checks-compiled-code/index.Rmd:103]: Even here I think capturing the `n` is a bit unusual. It's probably best practice, but I think few people do it, and you get to the heart of the problem faster without it.

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [inline content/blog/cran-checks-compiled-code/index.Rmd:152]:
```suggestion
It is a bit trickier if the destination is not a static buffer, so you'll have to determine the maximum `size` by carefully thinking about the code:
```

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [inline content/blog/cran-checks-compiled-code/index.Rmd:134]: That string is length 20, so there's no room for the null terminator?

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [reply content/blog/cran-checks-compiled-code/index.Rmd:152]: The example is better, thanks!

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd:103]: Fair enough. I was doing it to demonstrate the return value vs the `size` parameter - I've seen places where people capture it and compare it to the expected `size` and then condition on that...

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (jennybc) [reply content/blog/cran-checks-compiled-code/index.Rmd]: I'm not sure this readr/grisu thing is a great example because it's still basically static and, ideally, **we** would have captured the rationale for 32 bytes in **our** code. But we didn't.

Maybe @hadley is thinking more of an example where the size of the buffer depends on some data that can't be known in advance. This is cobbled together from reading from SO threads, so if you use it, make sure to get someone else to affirm it's not too dorky!

``` r
library(cpp11)
cpp_function('
  size_t say_hello(std::string name) {
    std::string fmt("Hello, %s!\\n");
    // subtract 2 for %s, add 1 for null terminator
    size_t size = fmt.size() - 2 + name.size() + 1;
    std::string buf;
    buf.resize(size);
    snprintf(&buf[0], size, fmt.c_str(), name.c_str());
    Rprintf(buf.c_str());
    return size;
  }
')
say_hello("Jenny")
#> Hello, Jenny!
#> [1] 15
```

<sup>Created on 2023-03-14 with [reprex v2.0.2.9000](https://reprex.tidyverse.org)</sup>

Comment (hadley) [reply content/blog/cran-checks-compiled-code/index.Rmd]: @jennybc I don't love that code either, because if you're using C++, I think you'd generally do something other than `sprintf()`. I can see places where we might use code like that, but it feels a bit artificial.

Maybe we can just omit an example for the case of a dynamic buffer since it will depend so much on the specific scenario.

Comment (jennybc) [reply content/blog/cran-checks-compiled-code/index.Rmd]: I searched r-lib and tidyverse packages for `snprintf()` and we really don't have any organic examples of a true dynamic buffer.

Comment (jennybc) [reply content/blog/cran-checks-compiled-code/index.Rmd]: @ateucher If we're going to keep the link and reference to the readr/grisu thing, I'd like the chance to slightly rephrase the text.

Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd]: I'm not at all opposed to taking out the dynamic buffer example - I don't think the intent is to show people _how_ to figure out the size of something, just to let them know that they need to with this change. It's also at risk of just getting too long...

@jennybc I'll leave it up to you if we keep it in - and definitely please go ahead and edit it if you think it's worth it. Though I think you have a good point about doing some gymnastics so as to not refactor it properly, so maybe it's not the best illustration?

Comment (jennybc) [reply content/blog/cran-checks-compiled-code/index.Rmd]: Let's just drop it, because it's such a mixed bag.

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd:103]: If you're ok with it @hadley I'd rather keep the `n` - I think it helps to demonstrate the slightly different return values of `sprintf()` vs `snprintf()`.

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd:18]: Should we try to do this for this post? The icons are here: https://github.com/tidyverse/tidyverse.org/tree/main/themes/hugo-graphite/static/css/images and the categories are defined here: https://github.com/tidyverse/tidyverse.org/blob/72bb2dd7ca0b2f211498a891aa54f55ddcad5014/themes/hugo-graphite/static/css/main-site.css#L835-L856

File: content/blog/cran-checks-compiled-code/index.Rmd
Comment (hadley) [reply content/blog/cran-checks-compiled-code/index.Rmd:103]: I don't have strong feelings about it so go ahead.

Comment (ateucher) [reply content/blog/cran-checks-compiled-code/index.Rmd:103]: Thanks!

**Pull request merged**
